### PR TITLE
Implement of GEP-995

### PR DESCRIPTION
**Pull request by howardjohn**
**What type of PR is this?**

/kind gep

**What this PR does / why we need it**:

Implements GEP-995

GEP: https://github.com/kubernetes-sigs/gateway-api/pull/2593 (merged)
Docs: https://github.com/kubernetes-sigs/gateway-api/pull/2946 (merged)

**Which issue(s) this PR fixes**:

Fixes #

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```
Release note was in 2593

### CODE REVIEW

**Review by LiorLieberman — COMMENTED** [review]
Generally looks good to me, 
One nit: the GEP [mentions](https://github.com/kubernetes-sigs/gateway-api/blame/7f9f51098f405344a3a7eb13c55895c77919f706/geps/gep-995/index.md#L56) 

"This format for the `name` field of route rules differs from the pattern adopted for the [`SectionName`](https://github.com/kubernetes-sigs/gateway-api/blob/f544a46ef92b7f234ee3e7bf50da35b05f862c35/apis/v1/shared_types.go#L624C10-L624C10) type, which was thought for specifying mainly DNS subdomain names ([RFC 1123](https://www.rfc-editor.org/rfc/rfc1123)), due to its use in the gateway listeners originally"

So although it is just a string, using v1.SectionName is a bit confusing. Maybe a dedicated type or simply a string?

**Comment by guicassolato** [comment]
> Generally looks good to me, One nit: the GEP [mentions](https://github.com/kubernetes-sigs/gateway-api/blame/7f9f51098f405344a3a7eb13c55895c77919f706/geps/gep-995/index.md#L56)
> 
> "This format for the `name` field of route rules differs from the pattern adopted for the [`SectionName`](https://github.com/kubernetes-sigs/gateway-api/blob/f544a46ef92b7f234ee3e7bf50da35b05f862c35/apis/v1/shared_types.go#L624C10-L624C10) type, which was thought for specifying mainly DNS subdomain names ([RFC 1123](https://www.rfc-editor.org/rfc/rfc1123)), due to its use in the gateway listeners originally"
> 
> So although it is just a string, using v1.SectionName is a bit confusing. Maybe a dedicated type or simply a string?

`SectionName` won't allow names such as `my-rule123`, `my_rule`, or `myRule`. https://go.dev/play/p/zRACGkGb_EG

Not sure what the best path is here, but I guess starting with `SectionName` is fine. Maybe it's the GEP that should change; then, if requested, another one opened to modify the definition of `SectionName` perhaps?

File: apis/v1/grpcroute_types.go [hunk 1]
```diff
@@ -143,13 +143,19 @@ type GRPCRouteSpec struct {
...
 }
 
 // GRPCRouteRule defines the semantics for matching a gRPC request based on
 // conditions (matches), processing it (filters), and forwarding the request to
 // an API object (backendRefs).
 type GRPCRouteRule struct {
+	// Name is the name of the route rule. This name MUST be unique within a Route if it is set.
+	//
+	// Support: Extended
+	Name SectionName `json:"name,omitempty"`
```
Comment (arkodg) [inline apis/v1/grpcroute_types.go:157]: optional tag and a ptr type, since this field is not mandatory ?

Comment (howardjohn) [reply apis/v1/grpcroute_types.go]: :+1: done

**Comment by robscott** [comment]
@guicassolato or @howardjohn I know this GEP is already essentially in scope for v1.2, but can one of you take a minute to leave a comment on https://github.com/kubernetes-sigs/gateway-api/discussions/3103 so we make sure we're tracking it?

**Comment by howardjohn** [comment]
Now that https://github.com/kubernetes-sigs/gateway-api/pull/3166 is merged I think this is unblocked. Let me know if there are any changes needed.

File: apis/v1/httproute_types.go [hunk 2]
```diff
@@ -117,6 +117,7 @@ type HTTPRouteSpec struct {
 	// Rules are a list of HTTP matchers, filters and actions.
 	//
 	// +optional
+	// +kubebuilder:validation:XValidation:message="Route name must be unique within the route",rule="self.all(l1, !has(l1.name) || self.exists_one(l2, has(l2.name) && l1.name == l2.name))"
```
Comment (robscott) [inline apis/v1/httproute_types.go:120]: Same comment applies throughout
```suggestion
	// +kubebuilder:validation:XValidation:message="Rule name must be unique within the route",rule="self.all(l1, !has(l1.name) || self.exists_one(l2, has(l2.name) && l1.name == l2.name))"
```

File: pkg/test/cel/httproute_test.go [hunk 3]
```diff
@@ -1209,6 +1209,18 @@ func TestHTTPRouteRule(t *testing.T) {
 				},
 			},
 		},
+		{
```
Comment (robscott) [inline pkg/test/cel/httproute_test.go:1212]: Can you add a valid test case here?

Comment (howardjohn) [reply pkg/test/cel/httproute_test.go:1212]: Yep, done. Included a "multiple unset is ok" part as well

**Comment by howardjohn** [comment]
@robscott can you take another look?

File: apis/v1/grpcroute_types.go [hunk 4]
```diff
@@ -144,13 +144,20 @@ type GRPCRouteSpec struct {
...
 
 // GRPCRouteRule defines the semantics for matching a gRPC request based on
 // conditions (matches), processing it (filters), and forwarding the request to
 // an API object (backendRefs).
 type GRPCRouteRule struct {
+	// Name is the name of the route rule. This name MUST be unique within a Route if it is set.
+	//
+	// Support: Extended
+	// +optional
+	Name *SectionName `json:"name,omitempty"`
```
Comment (robscott) [inline apis/v1/grpcroute_types.go:159]:
```suggestion
	<gateway:experimental>
	Name *SectionName `json:"name,omitempty"`
```

File: apis/v1/httproute_types.go [hunk 5]
```diff
@@ -133,6 +134,12 @@ type HTTPRouteSpec struct {
...
 // +kubebuilder:validation:XValidation:message="Within backendRefs, When using URLRewrite filter with path.replacePrefixMatch, exactly one PathPrefix match must be specified",rule="(has(self.backendRefs) && self.backendRefs.exists_one(b, (has(b.filters) && b.filters.exists_one(f, has(f.urlRewrite) && has(f.urlRewrite.path) && f.urlRewrite.path.type == 'ReplacePrefixMatch' && has(f.urlRewrite.path.replacePrefixMatch))) )) ? ((size(self.matches) != 1 || !has(self.matches[0].path) || self.matches[0].path.type != 'PathPrefix') ? false : true) : true"
 type HTTPRouteRule struct {
+	// Name is the name of the route rule. This name MUST be unique within a Route if it is set.
+	//
+	// Support: Extended
+	// +optional
+	Name *SectionName `json:"name,omitempty"`
```
Comment (robscott) [inline apis/v1/httproute_types.go:141]:
```suggestion
	<gateway:experimental>
	Name *SectionName `json:"name,omitempty"`
```

File: pkg/test/cel/httproute_test.go [hunk 6]
```diff
@@ -1251,6 +1251,29 @@ func TestHTTPRouteRule(t *testing.T) {
 				return rules
 			}(),
 		},
+		{
+			name:       "invalid because multiple names are repeated",
+			wantErrors: []string{"Rule name must be unique within the route"},
+			rules: []gatewayv1.HTTPRouteRule{
```
Comment (robscott) [inline pkg/test/cel/httproute_test.go:1257]: Mind adding another test case where the duplicate values are spread out in the list and not consecutive?

**Review by robscott — COMMENTED** [review]
Thanks @howardjohn!

File: apis/v1/grpcroute_types.go [hunk 7]
```diff
@@ -144,13 +144,21 @@ type GRPCRouteSpec struct {
...
+	// +kubebuilder:validation:XValidation:message="Rule name must be unique within the route",rule="self.all(l1, !has(l1.name) || self.exists_one(l2, has(l2.name) && l1.name == l2.name))"
```
Comment (robscott) [inline apis/v1/grpcroute_types.go:147]: Same suggestion applies throughout
```suggestion
	// +kubebuilder:experimental:validation:XValidation:message="Rule name must be unique within the route",rule="self.all(l1, !has(l1.name) || self.exists_one(l2, has(l2.name) && l1.name == l2.name))"
```

Relevant code is:
https://github.com/kubernetes-sigs/gateway-api/blob/d6e29e30272126964bf9b09f6b4688f3768ba657/pkg/generator/main.go#L177
https://github.com/kubernetes-sigs/gateway-api/blob/d6e29e30272126964bf9b09f6b4688f3768ba657/pkg/generator/main.go#L195

**Review by robscott — COMMENTED** [review]
Thanks @howardjohn! LGTM once tests pass.

/approve

File: apis/v1/grpcroute_types.go (hunk 7 above)
Comment (howardjohn) [reply apis/v1/grpcroute_types.go:147]: thanks!

**Review by mlavacca — COMMENTED** [review]
Thanks, @howardjohn!

/lgtm
